Sources/Contentful/Decodable.swift

```diff
@@ -0,0 +1,105 @@
+
+internal struct JSONCodingKeys: CodingKey {
+    internal var stringValue: String
+
+    internal init?(stringValue: String) {
+        self.stringValue = stringValue
+    }
+
+    internal var intValue: Int?
+
+    internal init?(intValue: Int) {
+        self.init(stringValue: "\(intValue)")
+        self.intValue = intValue
+    }
+}
+
+internal extension KeyedDecodingContainer {
+
+    internal func decode(_ type: Dictionary<String, Any>.Type, forKey key: K) throws -> Dictionary<String, Any> {
+        let container = try self.nestedContainer(keyedBy: JSONCodingKeys.self, forKey: key)
+        return try container.decode(type)
+    }
+
+    internal func decodeIfPresent(_ type: Dictionary<String, Any>.Type, forKey key: K) throws -> Dictionary<String, Any>? {
+        guard contains(key) else {
+            return nil
+        }
+        return try decode(type, forKey: key)
+    }
+
+    internal func decode(_ type: Array<Any>.Type, forKey key: K) throws -> Array<Any> {
+        var container = try self.nestedUnkeyedContainer(forKey: key)
+        return try container.decode(type)
+    }
+
+    internal func decodeIfPresent(_ type: Array<Any>.Type, forKey key: K) throws -> Array<Any>? {
+        guard contains(key) else {
+            return nil
+        }
+        return try decode(type, forKey: key)
+    }
+
+    internal func decode(_ type: Dictionary<String, Any>.Type) throws -> Dictionary<String, Any> {
+        var dictionary = Dictionary<String, Any>()
+
+        for key in allKeys {
+            if let intValue = try? decode(Int.self, forKey: key) {
+                dictionary[key.stringValue] = intValue
+            } else if let stringValue = try? decode(String.self, forKey: key) {
+                dictionary[key.stringValue] = stringValue
+            } else if let boolValue = try? decode(Bool.self, forKey: key) {
+                dictionary[key.stringValue] = boolValue
+            } else if let doubleValue = try? decode(Double.self, forKey: key) {
+                dictionary[key.stringValue] = doubleValue
+            } else if let fileMetaData = try? decode(Asset.FileMetadata.self, forKey: key) {
+                dictionary[key.stringValue] = fileMetaData // Custom contentful type.
+            } else if let nestedDictionary = try? decode(Dictionary<String, Any>.self, forKey: key) {
+                dictionary[key.stringValue] = nestedDictionary
+            } else if let nestedArray = try? decode(Array<Any>.self, forKey: key) {
+                dictionary[key.stringValue] = nestedArray
+            } else if try decodeNil(forKey: key) {
+                dictionary[key.stringValue] = true
+            }
+        }
+        return dictionary
+    }
+}
+
+internal extension UnkeyedDecodingContainer {
+
+    internal mutating func decode(_ type: Array<Any>.Type) throws -> Array<Any> {
+        var array: [Any] = []
+        while isAtEnd == false {
```

Comment:
I am failing to find `isAtEnd` is this a field defined somewhere?

Reply:
it's defined in the system class `UnkeyedDecodingContainer`